pylint/checkers/typecheck.py

```diff
@@ -523,6 +525,77 @@ def _emit_no_member(
     return True
 
 
+def _get_all_attribute_assignments(
+    node: astroid.FunctionDef, name: str | None = None
+) -> set[str]:
+    attributes = set()
+    for child in node.nodes_of_class((astroid.Assign, astroid.AnnAssign)):
+        targets = []
+        if isinstance(child, astroid.Assign):
+            targets = child.targets
+        elif isinstance(child, astroid.AnnAssign):
+            targets = [child.target]
+        for assign_target in targets:
+            if not isinstance(assign_target, astroid.AssignAttr):
+                continue
+            if not isinstance(assign_target.expr, astroid.Name):
+                continue
+            if name is None or assign_target.expr.name == name:
+                attributes.add(assign_target.attrname)
+    return attributes
+
+
+def _enum_has_attribute(
+    owner: astroid.Instance | nodes.ClassDef, node: nodes.Attribute
+) -> bool:
+    if isinstance(owner, astroid.Instance):
+        enum_def = next(
+            (b.parent for b in owner.bases if isinstance(b.parent, nodes.ClassDef)),
+            None,
+        )
+        if enum_def is None:
+            return False
+    else:
+        enum_def = owner
+    # Traverse the AST to find the Enum ClassDef
+    while not isinstance(enum_def, astroid.ClassDef):
+        enum_def = enum_def.parent
+
+    # Find __new__ and __init__
+    dunder_new = next((m for m in enum_def.methods() if m.name == "__new__"), None)
+    dunder_init = next((m for m in enum_def.methods() if m.name == "__init__"), None)
+    if dunder_new is None and dunder_init is None:
+        return False
```

Comment:
The last line is uncovered. I also wonder if we can do this without iterating over all methods. Doesn't `getattr` work?

Reply:
With getattr, I came up with something like this
```py
    dunder_new: astroid.FunctionDef | None = (enum_def.getattr("__new__") or [None])[0]
    dunder_init: astroid.FunctionDef | None = (enum_def.getattr("__init__") or [None])[0]
```

In practice, I'd expect there to always be an `__init__` or `__new__` on an enum, so we can probably drop the if None check (and the or `[None]` too)

Reply:
With getattr, the code could now explode with an uncaught `AttributeInferenceError` if someone were to write some truly awful code.

Reply:
Yeah I think that should be fine. Especially since we will always end up at `object`.

Reply:
The current way I've implemented it makes me feel more comfortable from a known types PoV

Reply:
Hmm okay. I'll keep this on my mental todo list. Iterating over all methods instead of doing something like `get()` seems like we could optimise this. Thanks for trying the other solutions, much appreciated 😄

Reply:
nw :)
And I agree that somelike like `.get(name, default=_, type=_)` would be quite nice, if no default is provided explode, if a type is provided return the type you expect, etc